### PR TITLE
feat: Move graceful shutdown code to pkgs, where it is missing

### PR DESCRIPTION
**Pull request by michaljurecko**
Jira: https://keboola.atlassian.net/browse/PSGO-597

**Changes:**
- Some componets had graceful shutdown solved externally (eg. `Stop` method).
- It's easier to test it, but it's easy to forget to call the method, that's why I moved graceful shutdown directly to components.

-----------

### CODE REVIEW

File: internal/pkg/service/stream/source/httpsource/httpsource_test.go [hunk 1]
```diff
@@ -58,7 +58,7 @@ func TestStart(t *testing.T) {
 	require.NoError(t, stream.StartComponents(ctx, d, mock.TestConfig(), stream.ComponentHTTPSource))
 
 	// Wait for the HTTP server
-	require.NoError(t, netutils.WaitForTCP(listenAddr, time.Second))
+	require.NoError(t, netutils.WaitForHTTP(url, 10*time.Second))
 	logger.AssertJSONMessages(t, `
 {"level":"info","message":"starting HTTP source node","component":"http-source"}
 {"level":"info","message":"started HTTP source on \"0.0.0.0:%d\"","component":"http-source"}
```
Comment (michaljurecko) [inline internal/pkg/service/stream/source/httpsource/httpsource_test.go:64]: Previous wait mechanism is not enough, for an unknown reason, maybe some `fasthttp` internal thing.

For standard Go HTTP server, the mechanism is OK: https://stackoverflow.com/a/53332769

image

File: internal/pkg/service/stream/storage/level/local/reader/volume/volumes.go [hunk 2]
```diff
@@ -6,19 +6,19 @@ import (
...
 )
 
-type collection = volume.Collection[*Volume]
-
 type Volumes struct {
-	*collection
-	logger log.Logger
+	logger     log.Logger
+	collection *volume.Collection[*Volume]
 }
```
Comment (michaljurecko) [inline internal/pkg/service/stream/storage/level/local/reader/volume/volumes.go:18]: Added `Collection` method, reader volumes now match writer volumes.

File: test/stream/api/base/api-startup/expected-server-stdout [hunk 3]
```diff
@@ -23,13 +23,13 @@
...
+{"level":"info","message":"closed etcd session","component":"volumes.registry.etcd.session"}
+{"level":"info","message":"stopped volumes registration","component":"volumes.registry"}
 {"level":"info","message":"closing volumes","component":"storage.node.writer"}
 {"level":"info","message":"closing volume","volume.path":"%s/hdd/my-volume","volume.id":"my-volume","component":"storage.node.writer.volume"}
 {"level":"info","message":"closed volume","volume.path":"%s/hdd/my-volume","volume.id":"my-volume","component":"storage.node.writer.volume"}
-{"level":"info","message":"received shutdown request","component":"volumes.registry"}
-{"level":"info","message":"closing etcd session: context canceled","component":"volumes.registry.etcd.session"}
-{"level":"info","message":"closed etcd session","component":"volumes.registry.etcd.session"}
-{"level":"info","message":"shutdown done","component":"volumes.registry"}
 {"level":"info","message":"closing volumes stream","component":"volume.repository"}
```
Comment (michaljurecko) [inline test/stream/api/base/api-startup/expected-server-stdout:33]: Fixed order issue, at first stop registration (let know other nodes, the volumes are no more available), then close volumes.

File: internal/pkg/service/stream/storage/statistics/collector/collector.go [hunk 4]
```diff
@@ -44,20 +42,32 @@ type writerSnapshot struct {
...
-	// Create context for cancellation of the periodical sync
-	var ctx context.Context
-	ctx, c.cancel = context.WithCancel(context.Background())
+	// Graceful shutdown
+	ctx, cancel := context.WithCancel(context.Background())
+	d.Process().OnShutdown(func(ctx context.Context) {
+		cancel()
+		c.logger.Info(ctx, "stopping storage statistics collector")
+		c.wg.Wait()
+		c.logger.Info(ctx, "storage statistics stopped")
```
Comment (Matovidlo) [inline internal/pkg/service/stream/storage/statistics/collector/collector.go:69]:
```suggestion
		c.logger.Info(ctx, "stopping storage statistics collector")
		defer c.logger.Info(ctx, "storage statistics stopped")
		cancel()
		c.wg.Wait()
```

**Review by Matovidlo — APPROVED** [review]
LGTM

**Review by hosekpeter — APPROVED** [review]
LGTM

File: internal/pkg/service/stream/storage/statistics/collector/collector.go (hunk 4 above)
Comment (michaljurecko) [reply internal/pkg/service/stream/storage/statistics/collector/collector.go:69]: Why? :slightly_smiling_face:

**Pull request merged**